Archive/EulerFermat.lean

```diff
@@ -0,0 +1,169 @@
+Authors: Ralf Stephan
+-/
+import Mathlib
+
+/-!
+# a^n+1 is prime only if n is a power of two
+
+The fact that primes of the form `a^n+1` must have `n` a power of two was stated and
+the proof outlined in three sentences by Euler in the year 1738 [E026]. A proof can be
+found in Hardy & Wright, An introduction to the theory of numbers (5th ed., 1979),
+p.15, Theorem 17. In the proof below we followed Euler's thoughts.
+-/
+
+open Nat
+
+lemma simplify1 (a m' : ℕ) (ha : a > 0) :
+    a ^ (2 * m' + 1) + 1 + (a + 1) * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) =
+    a ^ (2 * m' + 3) + 1 :=
+  have h₁ (a b c d : ℕ): a + b + c + d - a = b + c + d := by omega
+  have h₂ (a b : ℕ) (h : b ≤ a): 1 + a - b + b = 1 + a := by omega
+  calc
+    a ^ (2 * m' + 1) + 1 + (a + 1) * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) =
+      a ^ (2 * m' + 1) + 1 + a * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) +
+      (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) := by
+      linarith
+    _ = a ^ (2 * m' + 1) + 1 + a * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) +
+      a ^ (2 * m' + 2) - a ^ (2 * m' + 1) := by
+      rw [← Nat.add_sub_assoc]
+      exact pow_le_pow_of_le_right ha (le_succ (2 * m' + 1))
+    _ = 1 + a * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) + a ^ (2 * m' + 2) := by
+      rw [h₁]
+    _ = 1 + (a * a ^ (2 * m' + 2) - a * a ^ (2 * m' + 1)) + a ^ (2 * m' + 2) := by
+      rw [Nat.mul_sub_left_distrib a (a ^ (2 * m' + 2)) (a ^ (2 * m' + 1))]
+    _ = 1 + (a ^ (2 * m' + 2 + 1) - a ^ (2 * m' + 1 + 1)) + a ^ (2 * m' + 2) := by
+      simp_rw [← Nat.pow_succ']
+    _ = 1 + (a ^ (2 * m' + 3) - a ^ (2 * m' + 2)) + a ^ (2 * m' + 2) := by
+      linarith
+    _ = 1 + a ^ (2 * m' + 3) - a ^ (2 * m' + 2) + a ^ (2 * m' + 2) := by
+      rw [← Nat.add_sub_assoc]
+      exact pow_le_pow_of_le_right ha (le_succ (2 * m' + 2))
+    _ = 1 + a ^ (2 * m' + 3) := by
+      apply h₂ (a ^ (2 * m' + 3)) (a ^ (2 * m' + 2))
+      exact pow_le_pow_of_le_right ha (le_succ (2 * m' + 2))
+    _ = a ^ (2 * m' + 3) + 1 := by
+      exact Nat.add_comm 1 (a ^ (2 * m' + 3))
+
+/-- "$a^{2m+1}+1$ can be divided by $a+1$" [Euler, E026]. Proof by induction. -/
+theorem H1 (a m : ℕ) (ha : a > 0) : (a + 1) ∣ a ^ (2 * m + 1) + 1 := by
+  have h₃ (a m' : ℕ) : a ^ (2 * m' + 2 + 1) = a ^ (2 * m' + 3) := by ring_nf
+  have hdvd (a m' : ℕ) : (a + 1) ∣ (a + 1) * (a ^ (2 * m' + 2) - a ^ (2 * m' + 1)) :=
+    dvd_mul_right (a + 1) (a ^ (2 * m' + 2) - a ^ (2 * m' + 1))
+  induction m with
+  | zero =>
+    exact Dvd.intro_left (Nat.pow a 0) rfl
+  | succ m' ih =>
+    rw [Nat.mul_succ, h₃ a m', ← simplify1 a m' ha]
+    exact dvd_add ih (hdvd a m')
+
+/-- "$a^p+1$ divides $a^{p(2m+1)}+1$" [Euler, E026]. Substitute $a$ for $a^p$ in the above. -/
+lemma H2 (a m p: ℕ) (ha : a > 0) : (a ^ p + 1) ∣ a ^ (p * (2 * m + 1)) + 1 := by
+  rw [pow_mul a p (2 * m + 1)]
+  exact H1 (a ^ p) m (pos_pow_of_pos p ha)
+
+/-- The odd part of `n > 0` is either 1, or is `∈ [1, n)`. -/
```

Comment:
I'm not good with brackets, but I think you have them swapped in `[1, n)`.

Reply:
Right.